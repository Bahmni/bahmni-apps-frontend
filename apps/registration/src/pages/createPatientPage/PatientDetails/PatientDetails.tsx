import { Tile } from '@bahmni-frontend/bahmni-design-system';
import { useTranslation } from '@bahmni-frontend/bahmni-services';
import type { AddressHierarchyEntry } from '@bahmni-frontend/bahmni-services';
import type { AddressErrors } from '../../../models/address';
import type { PatientFormData } from '../../../models/patientForm';
import type {
  NameErrors,
  ValidationErrors,
  AgeErrors,
  DateErrors,
} from '../../../models/validation';
import { AdditionalInformationSection } from '../../../sections/AdditionalInformationSection';
import { AddressInformationSection } from '../../../sections/AddressInformationSection';
import { BasicInformationSection } from '../../../sections/BasicInformationSection';
import { ContactInformationSection } from '../../../sections/ContactInformationSection';
import { FormActions } from '../../../sections/FormActions';
import styles from './styles/index.module.scss';

interface PatientDetailsProps {
  // Form data
  formData: PatientFormData;
  identifierPrefixes: string[];
  genders: string[];

  // Error states
  nameErrors: NameErrors;
  validationErrors: ValidationErrors;
  ageErrors: AgeErrors;
  dateErrors: DateErrors;
  addressErrors: AddressErrors;

  // DOB states
  dobEstimated: boolean;

  // Address hierarchy states
  suggestions: {
    district: AddressHierarchyEntry[];
    state: AddressHierarchyEntry[];
    pincode: AddressHierarchyEntry[];
  };
  showSuggestions: {
    district: boolean;
    state: boolean;
    pincode: boolean;
  };

  // Mutation state
  isSaving: boolean;

  // Handlers
  onInputChange: (
    field: keyof PatientFormData,
    value: string | boolean,
  ) => void;
  onNameChange: (
    field: 'firstName' | 'middleName' | 'lastName',
    value: string,
  ) => void;
  onAgeChange: (
    field: 'ageYears' | 'ageMonths' | 'ageDays',
    value: string,
  ) => void;
  onDateInputChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onDateOfBirthChange: (dates: Date[]) => void;
  onDobEstimatedChange: (estimated: boolean) => void;
  onGenderChange: (gender: string) => void;
  onAddressInputChange: (
    field: 'district' | 'state' | 'pincode',
    value: string,
    addressField: string,
  ) => void;
  onSuggestionSelect: (
    field: 'district' | 'state' | 'pincode',
    entry: AddressHierarchyEntry,
  ) => void;
  onSuggestionBlur: (field: 'district' | 'state' | 'pincode') => void;
  onSuggestionFocus: (field: 'district' | 'state' | 'pincode') => void;
  onPhoneChange: (
    field: 'phoneNumber' | 'altPhoneNumber',
    value: string,
  ) => void;
  onSave: () => void;
}

export const PatientDetails = ({
  formData,
  identifierPrefixes,
  genders,
  nameErrors,
  validationErrors,
  ageErrors,
  dateErrors,
  addressErrors,
  dobEstimated,
  suggestions,
  showSuggestions,
  isSaving,
  onInputChange,
  onNameChange,
  onAgeChange,
  onDateInputChange,
  onDateOfBirthChange,
  onDobEstimatedChange,
  onGenderChange,
  onAddressInputChange,
  onSuggestionSelect,
  onSuggestionBlur,
  onSuggestionFocus,
  onPhoneChange,
  onSave,
}: PatientDetailsProps) => {
  const { t } = useTranslation();

  return (
    <>
      <Tile className={styles.patientDetailsHeader}>
        <span className={styles.sectionTitle}>
          {t('CREATE_PATIENT_HEADER_TITLE')}
        </span>
      </Tile>

      <div className={styles.formContainer}>
        <BasicInformationSection
          formData={formData}
          identifierPrefixes={identifierPrefixes}
          genders={genders}
          nameErrors={nameErrors}
          validationErrors={validationErrors}
          ageErrors={ageErrors}
          dateErrors={dateErrors}
          dobEstimated={dobEstimated}
          onInputChange={
            onInputChange as (
              field: keyof PatientFormData,
              value: string | boolean,
            ) => void
          }
          onNameChange={onNameChange}
          onAgeChange={onAgeChange}
          onDateInputChange={onDateInputChange}
          onDateOfBirthChange={onDateOfBirthChange}
          onDobEstimatedChange={onDobEstimatedChange}
          onGenderChange={onGenderChange}
        />

        <AddressInformationSection
          formData={formData}
          suggestions={suggestions}
          showSuggestions={showSuggestions}
          addressErrors={addressErrors}
          onInputChange={
            onInputChange as (
              field: keyof PatientFormData,
              value: string,
            ) => void
          }
          onAddressInputChange={onAddressInputChange}
          onSuggestionSelect={onSuggestionSelect}
          onSuggestionBlur={onSuggestionBlur}
          onSuggestionFocus={onSuggestionFocus}
        />

        <ContactInformationSection
          formData={formData}
          onPhoneChange={onPhoneChange}
        />

        <AdditionalInformationSection
          formData={formData}
          onInputChange={
            onInputChange as (
              field: keyof PatientFormData,
              value: string,
            ) => void
          }
        />
      </div>

      <FormActions onSave={onSave} isSaving={isSaving} />
    </>
  );
};
