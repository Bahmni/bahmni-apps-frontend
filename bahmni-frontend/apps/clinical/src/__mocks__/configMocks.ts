import { DashboardConfig } from "@bahmni-frontend/bahmni-services";

export const validFullClinicalConfig = {
  patientInformation: {
    displayPatientIdentifiers: true,
    showPatientPhoto: true,
    additionalAttributes: ['caste', 'education', 'occupation'],
  },
  actions: [
    {
      name: 'Start Visit',
      url: '/openmrs/ws/rest/v1/visit',
      icon: 'fa fa-stethoscope',
      requiredPrivilege: 'Start Visit',
    },
    {
      name: 'Add Diagnosis',
      url: '/openmrs/ws/rest/v1/diagnosis',
      icon: 'fa fa-heartbeat',
      requiredPrivilege: 'Add Diagnosis',
    },
    {
      name: 'Record Allergy',
      url: '/openmrs/ws/rest/v1/allergy',
      icon: 'fa fa-exclamation-triangle',
      requiredPrivilege: 'Record Allergy',
    },
  ],
  dashboards: [
    {
      name: 'Patient Information',
      url: 'patient-information',
      requiredPrivileges: ['View Patient Information'],
      icon: 'fa fa-user',
      default: true,
    },
    {
      name: 'Conditions',
      url: 'conditions',
      requiredPrivileges: ['View Conditions'],
      icon: 'fa fa-heartbeat',
    },
    {
      name: 'Allergies',
      url: 'allergies',
      requiredPrivileges: ['View Allergies'],
      icon: 'fa fa-exclamation-triangle',
    },
  ],
  consultationPad: {
    allergyConceptMap: {
      medicationAllergenUuid: '162552AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA',
      foodAllergenUuid: '162553AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA',
      environmentalAllergenUuid: '162554AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA',
      allergyReactionUuid: '162555AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA',
    },
  },
};

export const validDashboardConfig: DashboardConfig = {
  sections: [
    {
      id: 'vitals',
      name: 'Vitals',
      icon: 'heartbeat',
      controls: [],
    },
    {
      id: 'medications',
      name: 'Medications',
      icon: 'pills',
      controls: [],
    },
  ],
};