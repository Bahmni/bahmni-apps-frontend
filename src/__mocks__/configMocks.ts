// src/__mocks__/configMocks.ts

// Happy Path Mocks
export const validFullConfig = {
  section: [
    {
      name: 'Patient Information',
      translationKey: 'DASHBOARD_TITLE_PATIENT_INFO_KEY',
      description: 'Displays patient demographics and identifiers',
      icon: 'fa fa-user',
      controls: [
        {
          type: 'patientInformation',
          translationKey: 'DASHBOARD_TITLE_PATIENT_DETAILS_KEY',
        },
      ],
    },
    {
      name: 'Conditions',
      translationKey: 'DASHBOARD_TITLE_CONDITIONS_KEY',
      description: 'Displays patient conditions and diagnoses',
      icon: 'fa fa-heartbeat',
      controls: [
        {
          type: 'diagnosis',
          translationKey: 'DASHBOARD_TITLE_DIAGNOSIS_KEY',
        },
      ],
    },
    {
      name: 'Allergies',
      translationKey: 'DASHBOARD_TITLE_ALLERGIES_KEY',
      description: 'Displays patient allergies',
      icon: 'fa fa-exclamation-triangle',
      controls: [
        {
          type: 'allergies',
          translationKey: 'DASHBOARD_TITLE_ALLERGIES_LIST_KEY',
        },
      ],
    },
  ],
};

export const minimalConfig = {
  section: [
    {
      name: 'Basic Information',
      translationKey: 'DASHBOARD_TITLE_BASIC_KEY',
      description: 'Basic patient information',
      icon: 'fa fa-info',
      controls: [
        {
          type: 'patientInformation',
        },
      ],
    },
  ],
};

export const mixedConfig = {
  section: [
    {
      name: 'Required Section',
      translationKey: 'DASHBOARD_TITLE_REQUIRED_KEY',
      description: 'Section with required fields only',
      icon: 'fa fa-check',
      controls: [
        {
          type: 'patientInformation',
        },
      ],
    },
    {
      name: 'Optional Section',
      translationKey: 'DASHBOARD_TITLE_OPTIONAL_KEY',
      description: 'Section with optional fields',
      icon: 'fa fa-plus',
      controls: [
        {
          type: 'diagnosis',
          translationKey: 'DASHBOARD_TITLE_DIAGNOSIS_OPTIONAL_KEY',
        },
      ],
    },
  ],
};

// Sad Path Mocks
export const invalidConfig = {
  // Missing required section array
  otherProperty: 'value',
};

export const emptyResponse = null;

export const malformedJsonResponse = '{invalid-json}';

// Edge Case Mocks
export const largeConfig = {
  section: generateLargeSections(50), // Generates 50 sections
};

export const allOptionalFieldsConfig = {
  section: [
    {
      name: 'Comprehensive Section',
      translationKey: 'DASHBOARD_TITLE_COMPREHENSIVE_KEY',
      description: 'Section with all possible controls',
      icon: 'fa fa-th-large',
      controls: [
        {
          type: 'patientInformation',
          translationKey: 'DASHBOARD_TITLE_PATIENT_INFO_KEY',
        },
        {
          type: 'allergies',
          translationKey: 'DASHBOARD_TITLE_ALLERGIES_KEY',
        },
        {
          type: 'diagnosis',
          translationKey: 'DASHBOARD_TITLE_DIAGNOSIS_KEY',
        },
        {
          type: 'formsV2React',
          translationKey: 'DASHBOARD_TITLE_FORMS_KEY',
        },
      ],
    },
  ],
};

// Helper function to generate large config
function generateLargeSections(count: number) {
  const sections = [];
  const controlTypes = [
    'patientInformation',
    'allergies',
    'formsV2React',
    'diagnosis',
  ];

  for (let i = 0; i < count; i++) {
    sections.push({
      name: `Section ${i}`,
      translationKey: `DASHBOARD_TITLE_SECTION_${i}_KEY`,
      description: `Description for section ${i}`,
      icon: 'fa fa-folder',
      controls: [
        {
          type: controlTypes[i % controlTypes.length],
          translationKey: `DASHBOARD_TITLE_CONTROL_${i}_KEY`,
        },
      ],
    });
  }

  return sections;
}
