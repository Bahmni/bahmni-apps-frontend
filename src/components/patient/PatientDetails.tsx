import React, { useMemo } from 'react';
import { usePatient } from '@hooks/usePatient';
import { InlineNotification, SkeletonText, Tile, Stack } from '@carbon/react';
import { formatPatientData } from '@services/patientService';

interface PatientDetailsProps {
  patientUUID: string;
}

const PatientDetails: React.FC<PatientDetailsProps> = ({ patientUUID }) => {
  const { patient, loading, error } = usePatient(patientUUID);

  // Format patient data using the service
  const formattedPatient = useMemo(() => {
    if (!patient) return null;
    return formatPatientData(patient);
  }, [patient]);

  if (loading) {
    return (
      <Tile>
        <SkeletonText
          heading
          width="100%"
          lineCount={5}
          data-testid="skeleton-loader"
        />
      </Tile>
    );
  }

  if (error) {
    return (
      <InlineNotification kind="error" title="Error" subtitle={error.message} />
    );
  }

  if (!patient || !formattedPatient) {
    return (
      <InlineNotification
        kind="info"
        title="No data"
        subtitle="Patient information not found"
      />
    );
  }

  return (
    <Stack gap={5}>
      <Tile>
        {formattedPatient.fullName && <h2>{formattedPatient.fullName}</h2>}
        <p>ID: {formattedPatient.id}</p>
        {formattedPatient.gender && <p>Gender: {formattedPatient.gender}</p>}
        {formattedPatient.birthDate && (
          <p>Birth Date: {formattedPatient.birthDate}</p>
        )}
        {formattedPatient.formattedAddress && (
          <p>Address: {formattedPatient.formattedAddress}</p>
        )}
        {formattedPatient.formattedContact && (
          <p>Contact: {formattedPatient.formattedContact}</p>
        )}
      </Tile>
    </Stack>
  );
};

export default PatientDetails;
