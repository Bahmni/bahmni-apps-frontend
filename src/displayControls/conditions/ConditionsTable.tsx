import React, { useMemo } from 'react';
import { Tag } from '@carbon/react';
import { useTranslation } from 'react-i18next';
import { ExpandableDataTable } from '@components/common/expandableDataTable/ExpandableDataTable';
import { usePatientUUID } from '@hooks/usePatientUUID';
import { useConditions } from '@hooks/useConditions';
import { formatConditions } from '@services/conditionService';
import { ConditionStatus, FormattedCondition } from '@types/condition';
import { formatDate, formatDateTime } from '@utils/date';
import { FormatDateResult } from '@types/date';

/**
 * Maps condition status to appropriate tag type
 * @param status - The condition status
 * @returns The tag type for the status
 */
const getStatusTagType = (status: ConditionStatus): 'green' | 'gray' => {
  switch (status) {
    case ConditionStatus.Active:
      return 'green';
    case ConditionStatus.Inactive:
    default:
      return 'gray';
  }
};

/**
 * Component to display patient conditions in a DataTable with expandable rows
 */
const ConditionsTable: React.FC = () => {
  const { t } = useTranslation();
  const patientUUID = usePatientUUID();
  const { conditions, loading, error } = useConditions(patientUUID);

  // Define table headers
  const headers = useMemo(
    () => [
      { key: 'display', header: t('CONDITION_LIST_CONDITION') },
      { key: 'status', header: t('CONDITION_LIST_STATUS') },
      { key: 'onsetDate', header: t('CONDITION_TABLE_ONSET_DATE') },
      { key: 'recorder', header: t('CONDITION_TABLE_PROVIDER') },
      { key: 'recordedDate', header: t('CONDITION_TABLE_RECORDED_DATE') },
    ],
    [t],
  );

  // Format conditions for display
  const formattedConditions = useMemo(() => {
    if (!conditions || conditions.length === 0) return [];
    return formatConditions(conditions);
  }, [conditions]);

  // Function to render cell content based on the cell ID
  const renderCell = (condition: FormattedCondition, cellId: string) => {
    switch (cellId) {
      case 'display':
        return condition.display;
      case 'status':
        return (
          <Tag type={getStatusTagType(condition.status)}>
            {condition.status == ConditionStatus.Active
              ? t('CONDITION_LIST_ACTIVE')
              : t('CONDITION_LIST_INACTIVE')}
          </Tag>
        );
      case 'onsetDate': {
        const onsetDate: FormatDateResult = formatDate(
          condition.onsetDate || '',
        );
        return onsetDate.formattedResult || t('CONDITION_TABLE_NOT_AVAILABLE');
      }
      case 'recorder':
        return condition.recorder || t('CONDITION_TABLE_NOT_AVAILABLE');
      case 'recordedDate': {
        const recordedDate: FormatDateResult = formatDateTime(
          condition.recordedDate || '',
        );
        return (
          recordedDate.formattedResult || t('CONDITION_TABLE_NOT_AVAILABLE')
        );
      }
    }
  };

  // Function to render expanded content for a condition
  const renderExpandedContent = (condition: FormattedCondition) => {
    if (condition.note && condition.note.length > 0) {
      return condition.note.map((note, index) => (
        <p style={{ padding: '0.5rem' }} key={index}>
          {note}
        </p>
      ));
    }
    return undefined;
  };

  return (
    <div data-testid="condition-table">
      <ExpandableDataTable
        tableTitle={t('CONDITION_LIST_DISPLAY_CONTROL_TITLE')}
        rows={formattedConditions}
        headers={headers}
        renderCell={renderCell}
        renderExpandedContent={renderExpandedContent}
        loading={loading}
        error={error}
        ariaLabel={t('CONDITION_LIST_DISPLAY_CONTROL_TITLE')}
        emptyStateMessage={t('CONDITION_LIST_NO_CONDITIONS')}
      />
    </div>
  );
};

export default ConditionsTable;
