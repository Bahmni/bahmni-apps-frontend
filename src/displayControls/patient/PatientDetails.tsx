import { SkeletonText } from '@carbon/react';
import React, { useMemo } from 'react';
import { useTranslation } from 'react-i18next';
import BahmniIcon from '@/components/common/bahmniIcon/BahmniIcon';
import { ICON_SIZE } from '@/constants/icon';
import { usePatient } from '@hooks/usePatient';
import { usePatientUUID } from '@hooks/usePatientUUID';
import { formatPatientData } from '@services/patientService';
import * as styles from './__styles__/PatientDetails.module.scss';

// TODO: Extract this as a PatientDetails Display Control Component
const PatientDetails: React.FC = () => {
  const { t } = useTranslation();
  const patientUUID: string | null = usePatientUUID();
  const { patient, loading, error } = usePatient(patientUUID);

  // Format patient data using the service
  const formattedPatient = useMemo(() => {
    if (!patient) return null;
    return formatPatientData(patient);
  }, [patient]);

  if (loading || error || !patient || !formattedPatient) {
    return (
      <div className={styles.skeletonContainer}>
        <SkeletonText
          heading
          width="20%"
          lineCount={2}
          data-testid="skeleton-loader"
        />
        <SkeletonText
          width="50%"
          lineCount={3}
          data-testid="skeleton-loade-subheader"
        />
      </div>
    );
  }

  const formatField = (value?: string | number | null) => value ?? null;

  const formattedIdentifiers = formattedPatient.identifiers.size
    ? Array.from(formattedPatient.identifiers.entries())
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        .map(([key, value]) => `${value}`)
        .filter(Boolean)
        .join(' | ')
    : null;

  const formattedGender = formatField(formattedPatient.gender);

  const formattedAge =
    formattedPatient.age?.years !== undefined
      ? `${formattedPatient.age.years} ${t('CLINICAL_YEARS_TRANSLATION_KEY', { count: formattedPatient.age.years })}, ${formattedPatient.age.months} ${t('CLINICAL_MONTHS_TRANSLATION_KEY', { count: formattedPatient.age.months })}, ${formattedPatient.age.days} ${t('CLINICAL_DAYS_TRANSLATION_KEY', { count: formattedPatient.age.days })}`
      : null;

  const details = [formattedAge, formatField(formattedPatient.birthDate)]
    .filter(Boolean)
    .join(' | ');

  return (
    <div className={styles.header}>
      {formattedPatient.fullName && (
        <p data-testid="patient-name" className={styles.patientName}>
          {formattedPatient.fullName}
        </p>
      )}
      <div className={styles.patientDetails}>
        <div className={styles.identifierAndGenderWrapper}>
          {formattedIdentifiers && (
            <p className={styles.detailsWithIcon}>
              <BahmniIcon id="id-card" name="fa-id-card" size={ICON_SIZE.SM} />
              <span>{formattedIdentifiers}</span>
            </p>
          )}
          {formattedGender && (
            <p className={styles.detailsWithIcon}>
              <BahmniIcon
                id="gender"
                name="fa-mars-stroke-up"
                size={ICON_SIZE.SM}
              />
              <span>{formattedGender}</span>
            </p>
          )}
        </div>
        {details && (
          <p className={styles.detailsWithIcon}>
            <BahmniIcon id="age" name="fa-cake-candles" size={ICON_SIZE.SM} />
            <span>{details}</span>
          </p>
        )}
      </div>
    </div>
  );
};

export default PatientDetails;
