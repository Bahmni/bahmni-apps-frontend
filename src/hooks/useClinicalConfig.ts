import { useContext } from 'react';
import { ClinicalConfigContext } from '@contexts/ClinicalConfigContext';
import { ClinicalConfigContextType } from '@types/config';

/**
 * Custom hook to access the config context
 * @returns The config context values including config, loading state, and error
 */
export const useClinicalConfig = (): ClinicalConfigContextType => {
  const context = useContext(ClinicalConfigContext);

  if (!context) {
    throw new Error(
      'useClinicalConfig must be used within a ClinicalConfigProvider',
    );
  }

  return context;
};
