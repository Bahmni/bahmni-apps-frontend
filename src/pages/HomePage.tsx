import React, { Suspense } from 'react';
import { Grid, Column, Section, Loading } from '@carbon/react';
import { useConfig } from '@hooks/useConfig';

const PatientDetails = React.lazy(
  () => import('@components/patient/PatientDetails'),
);
const ConditionsTable = React.lazy(
  () => import('@components/conditions/ConditionsTable'),
);
const AllergiesTable = React.lazy(
  () => import('@components/allergies/AllergiesTable'),
);

const HomePage: React.FC = () => {
  const { config } = useConfig();

  if (!config) {
    return <Loading description="Loading..." />;
  }

  return (
    <Suspense fallback={<Loading description="Loading..." />}>
      <Section>
        <Grid>
          <Column lg={16} md={8} sm={4}>
            <PatientDetails />
            <AllergiesTable />
            <ConditionsTable />
          </Column>
        </Grid>
      </Section>
    </Suspense>
  );
};

export default HomePage;
