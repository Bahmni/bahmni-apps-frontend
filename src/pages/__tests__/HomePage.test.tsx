import React from 'react';
import { render, screen } from '@testing-library/react';
import HomePage from '../HomePage';
import PatientDetails from '@components/patient/PatientDetails';
import ConditionsTable from '@components/conditions/ConditionsTable';
import { useClinicalConfig } from '@hooks/useClinicalConfig';
import { validFullClinicalConfig } from '@__mocks__/configMocks';

// Mock Carbon components
jest.mock('@carbon/react', () => ({
  Grid: jest.fn(({ children }) => (
    <div data-testid="carbon-grid">{children}</div>
  )),
  Column: jest.fn(({ children }) => (
    <div data-testid="carbon-column">{children}</div>
  )),
  Section: jest.fn(({ children }) => (
    <div data-testid="carbon-section">{children}</div>
  )),
  Loading: jest.fn(() => <div data-testid="carbon-loading">Loading...</div>),
}));

// Mock the PatientDetails component
jest.mock('@components/patient/PatientDetails', () => {
  return jest.fn(() => (
    <div data-testid="mocked-patient-details">Mocked PatientDetails</div>
  ));
});

// Mock the ConditionsTable component
jest.mock('@components/conditions/ConditionsTable', () => {
  return jest.fn(() => (
    <div data-testid="mocked-conditions-table">Mocked ConditionsTable</div>
  ));
});

// Mock the useClinicalConfig hook
jest.mock('@hooks/useClinicalConfig');

// Mock the AllergiesTable component
jest.mock('@components/allergies/AllergiesTable', () => {
  return jest.fn(() => (
    <div data-testid="mocked-allergy-table">Mocked AllergiesTable</div>
  ));
});
describe('HomePage Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.spyOn(console, 'log').mockImplementation(() => {});
  });

  it('should show loading state when config is null', () => {
    // Mock useClinicalConfig to return null config (loading state)
    (useClinicalConfig as jest.Mock).mockReturnValue({ clinicalConfig: null });

    render(<HomePage />);

    // Should show loading component
    expect(screen.getByTestId('carbon-loading')).toBeInTheDocument();

    // Should not render other components
    expect(
      screen.queryByTestId('mocked-patient-details'),
    ).not.toBeInTheDocument();
    expect(
      screen.queryByTestId('mocked-conditions-table'),
    ).not.toBeInTheDocument();
  });

  it('should render with correct Carbon layout structure when config is loaded', () => {
    // Mock useClinicalConfig to return config
    (useClinicalConfig as jest.Mock).mockReturnValue({
      clinicalConfig: validFullClinicalConfig,
    });

    render(<HomePage />);

    // Should render Carbon layout components
    expect(screen.getByTestId('carbon-section')).toBeInTheDocument();
    expect(screen.getByTestId('carbon-grid')).toBeInTheDocument();
    expect(screen.getByTestId('carbon-column')).toBeInTheDocument();
  });

  it('should render child components when config is loaded', () => {
    // Mock useClinicalConfig to return config
    (useClinicalConfig as jest.Mock).mockReturnValue({
      clinicalConfig: validFullClinicalConfig,
    });

    render(<HomePage />);

    // Should render child components
    expect(PatientDetails).toHaveBeenCalled();
    expect(screen.getByTestId('mocked-patient-details')).toBeInTheDocument();
    expect(screen.getByText('Mocked PatientDetails')).toBeInTheDocument();

    expect(ConditionsTable).toHaveBeenCalled();
    expect(screen.getByTestId('mocked-conditions-table')).toBeInTheDocument();
    expect(screen.getByText('Mocked ConditionsTable')).toBeInTheDocument();
  });

  it('should match snapshot when loading', () => {
    // Mock useClinicalConfig to return null config (loading state)
    (useClinicalConfig as jest.Mock).mockReturnValue({ clinicalConfig: null });

    const { asFragment } = render(<HomePage />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should match snapshot when loaded', () => {
    // Mock useClinicalConfig to return config
    (useClinicalConfig as jest.Mock).mockReturnValue({
      clinicalConfig: validFullClinicalConfig,
    });

    const { asFragment } = render(<HomePage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
