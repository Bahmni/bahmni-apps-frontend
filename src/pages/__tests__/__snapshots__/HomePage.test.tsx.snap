// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Component should match snapshot when loaded 1`] = `
<DocumentFragment>
  <div
    data-testid="carbon-section"
  >
    <div
      data-testid="carbon-grid"
    >
      <div
        data-testid="carbon-column"
      >
        <div
          data-testid="mocked-patient-details"
        >
          Mocked PatientDetails
        </div>
        <div
          data-testid="mocked-allergy-table"
        >
          Mocked AllergiesTable
        </div>
        <div
          data-testid="mocked-conditions-table"
        >
          Mocked ConditionsTable
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`HomePage Component should match snapshot when loading 1`] = `
<DocumentFragment>
  <div
    data-testid="carbon-loading"
  >
    Loading...
  </div>
</DocumentFragment>
`;
